fleetctl/fleetctl.go

```diff
@@ -1067,78 +1072,87 @@ func assertUnitState(name string, js job.JobState, out io.Writer) (ret bool) {
-		return 0
-	}
-	if conn == nil {
-		stderr("Got a nil connection")
-		return 0
+func tryWaitForSystemdActiveState(units []string, maxAttempts int) (err error) {
+	if maxAttempts <= -1 {
+		for _, name := range units {
+			stdout("Triggered unit %s start", name)
+		}
+		return nil
 	}
-	defer conn.Close()
 
-	// Get systemd state and check the state is active & loaded.
-	// If any unit in []units cannot be found, then return an error.
-	// That should be the only case of returning 1.
 	for _, uName := range units {
-		found, err := waitForSystemdActiveState(conn, uName)
-		if !found || err != nil {
+		err := waitForSystemdActiveState(uName)
+		if err != nil {
 			stderr("cannot find an active unit: %v", err)
-			return 1
+			return err
 		}
 	}
 
-	return 0
+	return nil
 }
 
 // waitForSystemdActiveState tries to assert that the given unit becomes
 // active, retrying periodically until the unit gets active. It will retry
-// up to maxAttempts.
-func waitForSystemdActiveState(conn *sd_dbus.Conn, unitName string) (found bool, err error) {
-	return func() (found bool, errWait error) {
+// up to 15 seconds.
+func waitForSystemdActiveState(unitName string) (err error) {
+	return func() error {
 		timeout := 15 * time.Second
 		alarm := time.After(timeout)
 		ticker := time.Tick(250 * time.Millisecond)
 		for {
 			select {
 			case <-alarm:
-				return false, fmt.Errorf("Failed to reach expected state within %v.", timeout)
+				return fmt.Errorf("Failed to reach expected state within %v.", timeout)
 			case <-ticker:
-				if found, errA := assertSystemdActiveState(conn, unitName); found && errA == nil {
-					return found, nil
+				if errA := assertSystemdActiveState(unitName); errA == nil {
+					return nil
 				}
 			}
 		}
 	}()
 }
 
 // assertSystemdActiveState determines if a given systemd unit is actually
-// in the active state, making use of systemd's DBUS connection.
-func assertSystemdActiveState(conn *sd_dbus.Conn, unitName string) (found bool, err error) {
-	listUnits, err := conn.ListUnits()
+// in the active state, making use of cAPI
+func assertSystemdActiveState(unitName string) (err error) {
+	uState, err := getSingleUnitState(unitName)
 	if err != nil {
-		return false, fmt.Errorf("Failed to list systemd unit: %v", err)
+		return err
 	}
 
-	for _, u := range listUnits {
-		if strings.Compare(u.Name, unitName) != 0 {
-			continue
-		}
-		if u.ActiveState != "active" || u.LoadState != "loaded" {
-			return false, fmt.Errorf("Failed to find an active unit %s", unitName)
-		}
+	// Get systemd state and check the state is active & loaded.
+	if uState.ActiveState != "active" || uState.LoadState != "loaded" {
+		return fmt.Errorf("Failed to find an active unit %s", unitName)
+	}
+
+	return nil
+}
+
+// getSingleUnitState returns a single uState of type suState, which consists
+// of necessary systemd states, only for a given unit name.
+func getSingleUnitState(unitName string) (suState, error) {
```

Comment:
@dongsupark ah! good point, sorry about that. But now you've pointed that out - wouldn't it make more sense to unify the behaviour of the two so that they're both parallelised? :-)